### PR TITLE
Restrict editor programs

### PR DESCRIPTION
**Pull request by zlavergne**
**Requires a db migration**
This feature allows for project managers to restrict the editor programs that can be used during either editing stage.

* Added `map_editors` and `validate_editors` columns and accompanying dto's

* Added list of editors in the `Settings` menu of a project-edit page where a project manager can select permitted editors for each editing stage
Screen Shot 2019-03-21 at 10 37 50 AM

* Updated the editor dropdown for to populate only with the permitted editors
Screen Shot 2019-03-25 at 12 28 35 PM

> Note: iD X is added as an editor option. iD X will be the iD editor to support FB's ML as a Service project. It will be implemented in a future commit.

* only `ADMIN`s can add `iD X` as an editor program

* only user with `expertMode` enabled can use the `iD X` editor

### CODE REVIEW

**Comment by zlavergne** [comment]
Reset the branch in order to remove `id_x` implementation

**Comment by gaoxm** [comment]
@zlavergne the changes look good overall! I have a couple of ideas about the eng design (that may help simplify the changes) for discussion:
(1) I feel OK with opening iD X to both mappers and validators at the same time, as long as they're in expert mode. So the separation of editor lists between different roles may not be necessary.
(2) How much work will it be for TM admin to do the DB migration, given that they'll need to populate the value for all existing projects? If my assumption for (1) is true, then is it possible to just add a "id_x_enabled=true" type of metadata to some existing project column that contains a JSON blob? No DB migration will be needed that way, and we can default it to false when the property is not there (e.g. for the existing projects) :)

@jeffdefacto @chrisklaiber what do you guys think?

**Comment by zlavergne** [comment]
> (1) I feel OK with opening iD X to both mappers and validators at the same time, as long as they're in expert mode. So the separation of editor lists between different roles may not be necessary.

I did it this way after [this discussion on Slack](https://hotosm.slack.com/archives/CH19F91EU/p1553103960011100) with @jeffdefacto and Felix. Let me know if we want to go a different route.

> (2) How much work will it be for TM admin to do the DB migration, given that they'll need to populate the value for all existing projects? If my assumption for (1) is true, then is it possible to just add a "id_x_enabled=true" type of metadata to some existing project column that contains a JSON blob? No DB migration will be needed that way, and we can default it to false when the property is not there (e.g. for the existing projects) :)

Good catch. I was working with a fresh db and hadn't thought about that. I updated the migrations in [this commit](https://github.com/KaartGroup/tasking-manager/pull/14/commits/f1af4a0d3eaa4d87d9870a2716cdc69180009e86) to take this into account [with these lines.](https://github.com/KaartGroup/tasking-manager/blob/f1af4a0d3eaa4d87d9870a2716cdc69180009e86/migrations/versions/d01c4ceddeb4_.py#L24-L25) Now when alembic does the migration, it will add all the regular editors so they don't lose any functionality.

**Comment by zlavergne** [comment]
Circle is failing due to the debian version. See [#1141 on hotosm](https://github.com/hotosm/tasking-manager/pull/1411)

**Comment by jeffdefacto** [comment]
> (1) I feel OK with opening iD X to both mappers and validators at the same time, as long as they're in expert mode. So the separation of editor lists between different roles may not be necessary.

I think it makes sense to have them as a separate option. The feature is expanding upon only allowing JOSM for validation so I think to remain consistent we should allow the turning on or off of any editor for validation. For our workflow we don't use iD for the validation stage when using the service so turning it off for simplicity could be nice.

**Comment by gaoxm** [comment]
> > (1) I feel OK with opening iD X to both mappers and validators at the same time, as long as they're in expert mode. So the separation of editor lists between different roles may not be necessary.
> 
> I did it this way after [this discussion on Slack](https://hotosm.slack.com/archives/CH19F91EU/p1553103960011100) with @jeffdefacto and Felix. Let me know if we want to go a different route.
> 
> > (2) How much work will it be for TM admin to do the DB migration, given that they'll need to populate the value for all existing projects? If my assumption for (1) is true, then is it possible to just add a "id_x_enabled=true" type of metadata to some existing project column that contains a JSON blob? No DB migration will be needed that way, and we can default it to false when the property is not there (e.g. for the existing projects) :)
> 
> Good catch. I was working with a fresh db and hadn't thought about that. I updated the migrations in [this commit](https://github.com/KaartGroup/tasking-manager/pull/14/commits/f1af4a0d3eaa4d87d9870a2716cdc69180009e86) to take this into account [with these lines.](https://github.com/KaartGroup/tasking-manager/blob/f1af4a0d3eaa4d87d9870a2716cdc69180009e86/migrations/versions/d01c4ceddeb4_.py#L24-L25) Now when alembic does the migration, it will add all the regular editors so they don't lose any functionality.

Sorry I missed the earlier discussion on Slack. Will review the updates ASAP.

File: client/app/admin/edit-project/edit-project.controller.js [hunk 1]
```diff
@@ -40,6 +40,22 @@
             other: false
         };
 
+        // Editors for mapping
+        vm.mapEditors = {
```
Comment (gaoxm) [inline client/app/admin/edit-project/edit-project.controller.js]: minor naming suggestion: maybe just use "mappingEditors"? "mapEditors" sounds more like a human role rather than a tool.

File: client/app/admin/edit-project/edit-project.controller.js [hunk 2]
```diff
@@ -40,6 +40,22 @@
...
+            id: false,
+            josm: false,
+            potlatch2: false,
+            fieldpapers: false
+        };
+
+        // Editors for validation
+        vm.validateEditors = {
```
Comment (gaoxm) [inline client/app/admin/edit-project/edit-project.controller.js]: similarly, use "validationEditors"?

File: client/app/admin/edit-project/edit-project.controller.js [hunk 3]
```diff
@@ -167,6 +187,8 @@
             // Prepare the data for sending to API by removing any locales with no fields
             if (!requiredFieldsMissing && vm.editForm.$valid){
                 vm.project.mappingTypes = getMappingTypesArray();
+                vm.project.mapEditors = getMapEditorsArray();
+                vm.project.validateEditors = getValidateEditorsArray();
```
Comment (gaoxm) [inline client/app/admin/edit-project/edit-project.controller.js]: same

File: client/app/admin/edit-project/edit-project.controller.js [hunk 4]
```diff
@@ -928,6 +952,89 @@
...
 
+        /**
+         * Populate the mapping editor fields by checking which tags exist
+         * in the mapEditors array on the project
+         */
+        function populateEditorsForMapping(){
+            if (vm.project.mapEditors) {
+                if (vm.project.mapEditors.indexOf("ID") != -1) {
+                    vm.mapEditors.id = true;
+                }
```
Comment (gaoxm) [inline client/app/admin/edit-project/edit-project.controller.js]: maybe a little simpler this way?
  `vm.mapEditors.id = vm.project.mapEditors.indexOf("ID") != -1;`

File: client/app/admin/edit-project/edit-project.controller.js [hunk 5]
```diff
@@ -928,6 +952,89 @@
...
+                if (vm.project.mapEditors.indexOf("FIELD_PAPERS") != -1) {
+                    vm.mapEditors.fieldpapers = true;
+                }
+            }
+        }
+
+        /**
+         * Get map editors in array
+         */
+        function getMapEditorsArray(){
```
Comment (gaoxm) [inline client/app/admin/edit-project/edit-project.controller.js]: similarly, getMappingEditorsArray()

File: client/app/admin/edit-project/edit-project.controller.js [hunk 6]
```diff
@@ -928,6 +952,89 @@
...
+                if (vm.project.validateEditors.indexOf("FIELD_PAPERS") != -1) {
+                    vm.validateEditors.fieldpapers = true;
+                }
+            }
+        }
+
+        /**
+         * Get validate editors in array
+         */
+        function getValidateEditorsArray(){
```
Comment (gaoxm) [inline client/app/admin/edit-project/edit-project.controller.js]: getValidationEditorsArray()

File: client/app/admin/edit-project/edit-project.controller.js [hunk 7]
```diff
@@ -928,6 +952,89 @@
...
+
+
+        /**
+         * Populate the validation editor fields by checking which tags exist
+         * in the validateEditors array on the project
+         */
+        function populateEditorsForValidation(){
+            if (vm.project.validateEditors) {
+                if (vm.project.validateEditors.indexOf("ID") != -1) {
+                    vm.validateEditors.id = true;
```
Comment (gaoxm) [inline client/app/admin/edit-project/edit-project.controller.js]: same

File: client/app/project/project.controller.js [hunk 8]
```diff
@@ -59,7 +59,10 @@
 
         //editor
         vm.editorStartError = '';
-        vm.selectedEditor = 'ideditor';
+        vm.selectedMapEditor = 'ideditor';
+        vm.selectedValidateEditor = 'josm';
```
Comment (gaoxm) [inline client/app/project/project.controller.js]: same: `selectedMappingEditor` and `selectedValidationEditor`

File: client/app/project/project.controller.js [hunk 9]
```diff
@@ -115,6 +119,7 @@
                 var resultsPromise = accountService.getUser(session.username);
                 resultsPromise.then(function (user) {
                     vm.user = user;
+                    console.log(user);
```
Comment (gaoxm) [inline client/app/project/project.controller.js]: remove

File: client/app/services/editor.service.js [hunk 10]
```diff
@@ -79,6 +80,40 @@
             $window.open(url);
         }
 
+        /**
+         * Launch the iD X editor
```
Comment (jeffdefacto) [inline client/app/services/editor.service.js]: Shouldn't the id X stuff be a separate PR?

File: migrations/versions/215392b11b0e_.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_index('idx_project_info composite', 'project_info', ['locale', 'project_id'], unique=False)
+    op.create_index('textsearch_idx', 'project_info', ['text_searchable'], unique=False)
+    op.add_column('projects', sa.Column('map_editors', postgresql.ARRAY(sa.Integer()), nullable=True))
+    op.add_column('projects', sa.Column('validate_editors', postgresql.ARRAY(sa.Integer()), nullable=True))
```
Comment (gaoxm) [inline migrations/versions/215392b11b0e_.py]: same: `mapping_editors` and `validation_editors`

File: client/app/services/editor.service.js (hunk 10 above)
Comment (zlavergne) [reply client/app/services/editor.service.js]: Missed that one. Fixed with 29d203a

File: migrations/versions/215392b11b0e_.py [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+    op.alter_column('projects', 'task_creation_mode',
+               existing_type=sa.INTEGER(),
+               nullable=False)
+    op.create_index(op.f('ix_projects_map_editors'), 'projects', ['map_editors'], unique=False)
+    op.create_index(op.f('ix_projects_validate_editors'), 'projects', ['validate_editors'], unique=False)
```
Comment (gaoxm) [inline migrations/versions/215392b11b0e_.py]: do we need indices on these columns? For use cases where you want to search for all projects with "iD X" enabled, does this index support that?

File: migrations/versions/d01c4ceddeb4_.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+down_revision = 'df617df60678'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.drop_index('idx_project_info composite', table_name='project_info')
+    op.create_index('idx_project_info composite', 'project_info', ['locale', 'project_id'], unique=False)
+    op.drop_index('idx_project_info composite', table_name='project_info')
```
Comment (gaoxm) [inline migrations/versions/d01c4ceddeb4_.py]: why is it dropped again right after creation?

File: migrations/versions/d01c4ceddeb4_.py [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+    op.add_column('projects', sa.Column('map_editors', postgresql.ARRAY(sa.Integer()), nullable=False, server_default='{0,1,2,3}'))
+    op.add_column('projects', sa.Column('validate_editors', postgresql.ARRAY(sa.Integer()), nullable=False, server_default='{0,1,2,3}'))
+    op.create_index(op.f('ix_projects_map_editors'), 'projects', ['map_editors'], unique=False)
+    op.create_index(op.f('ix_projects_validate_editors'), 'projects', ['validate_editors'], unique=False)
```
Comment (gaoxm) [inline migrations/versions/d01c4ceddeb4_.py]: same questions

File: server/models/dtos/project_dto.py [hunk 15]
```diff
@@ -105,6 +118,8 @@ class ProjectDTO(Model):
     active_mappers = IntType(serialized_name='activeMappers')
     task_creation_mode = StringType(required=True, serialized_name='taskCreationMode',
                                     validators=[is_known_task_creation_mode], serialize_when_none=False)
+    map_editors = ListType(StringType, serialized_name='mapEditors', validators=[is_known_editor])
+    validate_editors = ListType(StringType, serialized_name='validateEditors', validators=[is_known_editor])
```
Comment (gaoxm) [inline server/models/dtos/project_dto.py]: same: `mapping_editors` and `validation_editors`

File: server/models/dtos/project_dto.py [hunk 16]
```diff
@@ -118,6 +133,8 @@ class ProjectSearchDTO(Model):
     page = IntType(required=True)
     text_search = StringType()
     is_project_manager = BooleanType(required=True, default=False)
+    map_editors = ListType(StringType, validators=[is_known_editor])
+    validate_editors = ListType(StringType, validators=[is_known_editor])
```
Comment (gaoxm) [inline server/models/dtos/project_dto.py]: same

File: server/models/dtos/project_dto.py [hunk 17]
```diff
@@ -129,9 +146,17 @@ def __hash__(self):
         if self.project_statuses:
             for project_status in self.project_statuses:
                 hashable_project_statuses = hashable_project_statuses + project_status
+        hashable_map_editors = ''
+        if self.map_editors:
+            for map_editor in self.map_editors:
+                hashable_map_editors = hashable_map_editors + map_editor
+        hashable_validate_editors = ''
```
Comment (gaoxm) [inline server/models/dtos/project_dto.py]: same: `hashable_mapping_editors` and `hashable_validation_editors`

File: server/models/postgis/project.py [hunk 18]
```diff
@@ -64,6 +64,10 @@ class Project(db.Model):
     organisation_tag = db.Column(db.String, index=True)
     campaign_tag = db.Column(db.String, index=True)
 
+    # Editors
+    map_editors = db.Column(ARRAY(db.Integer), default=[0, 1, 2, 3], index=True, nullable=False)
+    validate_editors = db.Column(ARRAY(db.Integer), default=[0, 1, 2, 3], index=True, nullable=False)
```
Comment (gaoxm) [inline server/models/postgis/project.py]: same for naming; also, how about using the Editors Enum values below for setting the defaults?

File: server/models/postgis/project.py [hunk 19]
```diff
@@ -203,6 +207,17 @@ def update(self, project_dto: ProjectDTO):
             type_array.append(MappingTypes[mapping_type].value)
         self.mapping_types = type_array
 
+        # Cast Editor strings to int array
+        map_editors_array = []
+        for map_editor in project_dto.map_editors:
+            map_editors_array.append(Editors[map_editor].value)
+        self.map_editors = map_editors_array
+
+        validate_editors_array = []
```
Comment (gaoxm) [inline server/models/postgis/project.py]: `mapping_editors_array` and `validation_editors_array`

File: server/models/postgis/project.py [hunk 20]
```diff
@@ -369,6 +384,20 @@ def _get_project_and_base_dto(self):
...
+                map_editors.append(Editors(map_editor).name)
+
+            base_dto.map_editors = map_editors
+
+        if self.validate_editors:
+            validate_editors = []
+            for validate_editor in self.validate_editors:
+                validate_editors.append(Editors(validate_editor).name)
+
+            base_dto.validate_editors = validate_editors
```
Comment (gaoxm) [inline server/models/postgis/project.py]: same

**Review by gaoxm — CHANGES_REQUESTED** [review]
looks good overall! Some minor comments inline, mostly around naming

File: migrations/versions/215392b11b0e_.py (hunk 12 above)
Comment (zlavergne) [reply migrations/versions/215392b11b0e_.py]: Yes. I put that in so projects would be searchable based on editors in the future. I tried to mimic similar implementations.

Comment (gaoxm) [reply migrations/versions/215392b11b0e_.py]: 👍 sounds good!

File: migrations/versions/d01c4ceddeb4_.py (hunk 13 above)
Comment (zlavergne) [reply migrations/versions/d01c4ceddeb4_.py]: I'm not entirely sure. That part was autogenerated by alembic. I'll remove the `drop`s. I can't think of any reason they would need to be there.

File: server/models/postgis/project.py (hunk 18 above)
Comment (zlavergne) [reply server/models/postgis/project.py]: Do you mean just replacing the ints with `Editors[editor]` or making it an `ARRAY(Enum(Editors))`?

File: client/app/services/editor.service.js [hunk 21]
```diff
@@ -18,6 +18,7 @@
             launchFieldPapersEditor: launchFieldPapersEditor,
             launchPotlatch2Editor: launchPotlatch2Editor,
             launchIdEditor: launchIdEditor,
+            launchIdX: launchIdX,
```
Comment (jeffdefacto) [inline client/app/services/editor.service.js]: Here too :)

Comment (zlavergne) [reply client/app/services/editor.service.js]: Argggghhhh. Search and Replace has failed me miserably

File: server/models/postgis/project.py (hunk 18 above)
Comment (gaoxm) [reply server/models/postgis/project.py]: the former, e.g. default=[Editors.ID.value, ...]

File: client/app/services/editor.service.js (hunk 21 above)
Comment (zlavergne) [reply client/app/services/editor.service.js]: Fixed 23c4bb4

File: server/models/postgis/project.py (hunk 20 above)
Comment (zlavergne) [reply server/models/postgis/project.py]: Fixed 23c4bb4

File: server/models/postgis/project.py (hunk 19 above)
Comment (zlavergne) [reply server/models/postgis/project.py]: Fixed 23c4bb4

File: server/models/postgis/project.py (hunk 18 above)
Comment (zlavergne) [reply server/models/postgis/project.py]: Fixed 23c4bb4

File: server/models/dtos/project_dto.py (hunk 17 above)
Comment (zlavergne) [reply server/models/dtos/project_dto.py]: Fixed 23c4bb4

**Comment by zlavergne** [comment]
23c4bb4 fixes all review changes

**Review by gaoxm — APPROVED** [review]

**Comment by gaoxm** [comment]
LGTM :)

**Review by jeffdefacto — APPROVED** [review]

**Comment by zlavergne** [comment]
Closing this since it was merged into mainline way back when and has been piped into ours with the sync to mainline.
